include/zephyr/sys/util.h

```diff
@@ -110,6 +110,44 @@ extern "C" {
 
 #endif /* __cplusplus */
 
+/**
+ * @brief Whether @p ptr is an element of @p array
+ *
+ * This macro can be seen as a slightly stricter version of @ref PART_OF_ARRAY
+ * in that it also ensures that @p ptr is aligned to an array-element boundary
+ * of @p array.
+ *
+ * In C, passing a pointer as @p array causes a compile error.
+ *
+ * @param array the array in question
+ * @param ptr the pointer to check
+ *
+ * @return 1 if @p ptr is part of @p array, 0 otherwise
+ */
+#define IS_ARRAY_ELEMENT(array, ptr)                                                               \
+	((uint8_t *)&(array)[0] <= (uint8_t *)(ptr) &&                                             \
+	 (uint8_t *)(ptr) < (uint8_t *)&(array)[ARRAY_SIZE(array)] &&                              \
+	 ((uint8_t *)(ptr) - (uint8_t *)(array)) % sizeof((array)[0]) == 0)
```

Comment:
> Could we consider to, if we add this, to deprecate `PART_OF_ARRAY` then? 

PART_OF_ARRAY is also useful since checks the whole range and not specific intervals.

> Alternatively, for `ARRAY_INDEX`, is it possible to floor the pointer given to that to be the head of the element, so that the `ptr` could be any field within a struct? 

That would explicitly break the use case for which I'm adding this macro (a thread stack pool, and allocator). 

> I'm just trying to see if we can add the `ARRAY_INDEX` without requiring a new `PART_OF_ARRAY` macro

It should be fairly trivial to come up with an `ARRAY_INDEX_FLOOR()` to do what you suggest. That could use `PART_OF_ARRAY()` instead and slightly different arithmetic.

Reply:
I am wondering if flooring shouldn't be the default. With the current PR, we would have the following
```c
void foo()
{
	struct my_inner_struct {
		int b;
	};

	struct my_struct {
		int a;
		struct my_inner_struct mis;
	};

	struct my_struct my_array[10] = { 0 };

	struct my_struct *my_elem = &my_array[5];

	uintptr_t my_index = ARRAY_INDEX(my_elem); /* 5*/

	struct my_inner_struct *my_inner_elem = &my_array[5].mis;

	my_index = ARRAY_INDEX(my_inner_elem); /* __ASSERT */
}
```

Would it really be better to have `ARRAY_INDEX(my_inner_elem)` throw an assert, than returning the index of the array element that it is a part of? 

The application could fairly easily use `CONTAINER_OF` on `my_inner_elem` to get the `struct my_struct` and then use it, so it's not a huge deal, but I am curious as to why you want to disallow doing `ARRAY_INDEX(my_inner_elem)` and what we gain from that restriction

Reply:
Well, for one, thread stacks would never properly be freed if we made the imprecise version the default.

There is a significantly easier way to get array index using a pointer to a struct member in an array of structs without CONTAINER_OF.

Reply:
> Well, for one, thread stacks would never properly be freed if we made the imprecise version the default.

Why not? 

> There is a significantly easier way to get array index using a pointer to a struct member in an array of structs without CONTAINER_OF.

Can you elaborate on what way that is?